File: modules/meeterm-terminal/ios/AttachmentInsertionPolicy.swift
import Foundation

/**
 * IME-safe attachment insertion decision.
 *
 * An active marked-text composition keeps the request `held` and is never
 * committed, cleared, or forwarded. Identity mismatches and Phase-A core
 * unavailability stay distinct so the sheet can explain each one.
 */
enum AttachmentInsertionPolicy {
  enum Verdict: Equatable {
    case heldComposing
    case heldNoAttachment
    case rejected(String, String)
    case readyToInsert
  }

  static func insert(
    composing: Bool,
    hasActiveSession: Bool,
    hasPreparedImage: Bool,
    hasUploadedPath: Bool,
    sessionTerminalId: String?,
    requestTerminalId: String
  ) -> Verdict {
    if composing { return .heldComposing }
    guard let sessionTerminalId = sessionTerminalId, hasActiveSession else {
      return .heldNoAttachment
    }
    guard sessionTerminalId == requestTerminalId else {
      return .rejected(
        AttachmentLimits.errorDestinationChanged,
        "The attachment belongs to a different terminal."
      )
    }
    if hasUploadedPath || hasPreparedImage { return .readyToInsert }
    return .rejected(
      AttachmentLimits.errorMissing,
      "The prepared image is missing; prepare it again."
    )
  }
}

File: modules/meeterm-terminal/android/src/main/java/dev/meeterm/terminal/AttachmentInsertionPolicy.kt
package dev.meeterm.terminal

/**
 * IME-safe attachment insertion decision.
 *
 * The composition check is first-class: an active Japanese conversion keeps
 * the request `held` and never commits, clears, or forwards preedit text.
 * Identity mismatches and Phase-A core unavailability are separate results so
 * the sheet can explain each one distinctly.
 */
internal object AttachmentInsertionPolicy {
  sealed class Verdict {
    data object HeldComposing : Verdict()
    data object HeldNoAttachment : Verdict()
    data class Rejected(val errorCode: String, val message: String) : Verdict()
    data object ReadyToInsert : Verdict()
  }

  fun begin(
    composing: Boolean,
    hasActiveSession: Boolean,
  ): Verdict = when {
    composing -> Verdict.HeldComposing
    hasActiveSession -> Verdict.ReadyToInsert
    else -> Verdict.Rejected(
      AttachmentLimits.ERROR_STATE,
      "No attachment session is active.",
    )
  }

  fun insert(
    composing: Boolean,
    hasActiveSession: Boolean,
    hasPreparedImage: Boolean,
    hasUploadedPath: Boolean,
    sessionTerminalId: String?,
    requestTerminalId: String,
  ): Verdict = when {
    composing -> Verdict.HeldComposing
    sessionTerminalId == null || !hasActiveSession -> Verdict.HeldNoAttachment
    sessionTerminalId != requestTerminalId -> Verdict.Rejected(
      AttachmentLimits.ERROR_TARGET,
      "The attachment belongs to a different terminal.",
    )
    hasUploadedPath || hasPreparedImage -> Verdict.ReadyToInsert
    else -> Verdict.Rejected(
      AttachmentLimits.ERROR_MISSING,
      "The prepared image is missing; prepare it again.",
    )
  }
}
